src/helpers/buffers/send.rs

```diff
@@ -49,35 +57,59 @@ impl SendBuffer {
         }
     }
 
-    pub fn push(&mut self, channel_id: &ChannelId, msg: &MessageEnvelope) -> Option<Vec<u8>> {
+    pub fn push(
+        &mut self,
+        channel_id: &ChannelId,
+        msg: &MessageEnvelope,
+        total_records: NonZeroUsize,
+    ) -> Option<Vec<u8>> {
         debug_assert!(
             msg.payload.len() <= ByteBuf::ELEMENT_SIZE_BYTES,
             "Message payload exceeds the maximum allowed size"
         );
 
-        let buf = if let Some(buf) = self.inner.get_mut(channel_id) {
-            buf
+        let channel = if let Some(channel) = self.inner.get_mut(channel_id) {
+            channel
         } else {
             self.inner.entry(channel_id.clone()).or_insert_with(|| {
+                tracing::trace!("create tx channel {:?}", channel_id);
                 let size =
                     NonZeroUsize::new(self.batch_count.get() * self.items_in_batch.get()).unwrap();
-                FixedSizeByteVec::new(size)
+                SendChannel {
+                    total_records: TotalRecords::Specified(total_records),
+                    sent_records: 0,
+                    buf: FixedSizeByteVec::new(size),
+                }
             })
         };
 
         // TODO: avoid the copy here and size the element size to the message type.
         let mut payload = [0; ByteBuf::ELEMENT_SIZE_BYTES];
         payload[..msg.payload.len()].copy_from_slice(&msg.payload);
-        buf.insert(channel_id, usize::from(msg.record_id), &payload);
-        buf.take(self.items_in_batch.get())
+        channel.buf.insert(channel_id, usize::from(msg.record_id), &payload);
```

Comment:
it would be great if `SendChannel` could insert items into its own buffer and update `send_records` counter by exposing an `insert` method. Then this code wouldn't need to worry about proper bookkeeping inside the send channel